gradio/routes.py

```diff
@@ -145,12 +145,16 @@ def get_blocks(self) -> gradio.Blocks:
             raise ValueError("No Blocks has been configured for this app.")
         return self.blocks
 
-    @staticmethod
-    def build_proxy_request(url_path):
+    def build_proxy_request(self, url_path):
         url = httpx.URL(url_path)
-        is_hf_url = url.host.endswith(".hf.space")
+        assert self.blocks
+        # Don't proxy a URL unless it's a URL specifically loaded by the user using
+        # gr.load() to prevent SSRF or harvesting of HF tokens by malicious Spaces.
+        is_safe_url = any(url.host.endswith(root) for root in self.blocks.root_urls)
+        if not is_safe_url:
+            raise ValueError("This URL cannot be proxied.")
```

Comment:
I think ValueError is a bit too generic, how about
```suggestion
            raise PermissionError("This URL cannot be proxied.")
```